guppylang/checker/expr_checker.py

```diff
@@ -544,7 +545,8 @@ def synthesize_instance_func(
                 f"expected `{exp_sig}`",
                 node,
             )
-        return func.synthesize_call([node, *args], node, self.ctx)
+        result, result_ty = func.synthesize_call([node, *args], node, self.ctx)
+        return with_type(result_ty, result), result_ty
```

Comment:
Drive-by: Annotate the AST node with the inferred type to avoid repeated checking